### ISSUE TITLE
Collection expressions: Incorrect diagnostic reported for element conversion error with CollectionBuilder type

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#69980]
**Version Used**: 

**Steps to Reproduce**:

```csharp
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

[CollectionBuilder(typeof(MyCollectionBuilder), "Create")]
class MyCollection<T> : IEnumerable<T>
{
    public IEnumerator<T> GetEnumerator() => default;
    IEnumerator IEnumerable.GetEnumerator() => default;
}

class MyCollectionBuilder
{
    public static MyCollection<T> Create<T>(ReadOnlySpan<T> items) => default;
}

class Program
{
    static void Main()
    {
        MyCollection<int> c = [1, 2, null];
    }
}
```

**Expected Behavior**:

```
(22,38): error CS1503: cannot convert from '<null>' to 'int'
```

**Actual Behavior**:

```
(22,31): error CS9188: 'MyCollection<int>' has a CollectionBuilderAttribute but no element type.
```
